[Chapter2 혼자 구현하기]

sezikim opened the pull request:
## 📌 관심있게 봐주셨으면 하는 사항
- TranSactionAnalyzer의 필드와 메소드가 잘 나뉘었는지 궁금합니다.
- CSVFileReader와 Transaction사이의 의존성을 낮출 수 있는 방법이 있을까요?
- 이외에 수정하면 좋은 부분 알려주시면 감사합니다.

sezikim committed:
프로젝트 생성

sezikim committed:
기능 구현사항 추가

sezikim committed:
feature. CSV 파일 Transaction 데이터로 저장하는 기능

sezikim committed:
feature. 입출금내역 분석, 출력 기능

sezikim committed:
README 수정

sezikim committed:
refactor TrasnSactionAnalyzer의 화면출력 기능 분리

sezikim committed:
chore. 상수 접근제어자, 사용하지 않는 상수 제거

minhee0327 commented on src/main/java/com/sezikim/bank/App.java:
```diff
@@ -0,0 +1,24 @@
+    public static void main(String[] args) {
+
+        try {
+            List<Transaction> transactionList = new CSVFileReader(CSV_FILE_PATH).readCSVFile();
```
hoxy.. 별건 아니지만.. 확장자를 (txt -> csv)변경해보는건 어떨까요..?

제 생각인데 나중에 txt, csv, excel, ... 어떤 파일로 받아도 해석할 수 있게끔 뭔가 인터페이스를 두는 방법도 좋을 거같아요!
요구사항에 없었지만 이부분도 고민이 되었던 부분인지라 🤗 ㅎㅎ

minhee0327 commented on src/main/java/com/sezikim/bank/App.java:
```diff
@@ -0,0 +1,24 @@
+
+public class App {
+
+    private static final String CSV_FILE_PATH = "src/main/resources/data.txt";
```
hoxy.. 별건 아니지만.. 확장자를 (txt -> csv)변경해보는건 어떨까요..?

- 제 생각인데 나중에 txt, csv, excel, ... 어떤 파일로 받아도 해석할 수 있게끔 뭔가 인터페이스를 두는 방법도 좋을 거같아요!
- 요구사항에 없었지만 이부분도 고민이 되었던 부분인지라 🤗 ㅎㅎ

sezikim replied:
앗... (요구사항은 사실.. 텍스트파일 입니다😃 )

minhee0327 commented on src/main/java/com/sezikim/bank/util/CSVFileReader.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    private Transaction createTransaction(String line) {
+        StringTokenizer stk = new StringTokenizer(line, ", ");
```
StringTokenizer를 사용한 방법 좋은거같아요! (저도 바꿔야겠군요..! ㅎㅎ)
대신 저는.. stk.hasNext()로 다음 토큰이 있는지 유무를 확인한 후에 date, value, category를 파싱해봐야겠어요!

sezikim replied:
stk.hasNext() 사용해서 유무 확인하는게 더 안전하겠네요!
감사합니다.

minhee0327 commented on src/main/java/com/sezikim/bank/TransactionAnalyzer.java:
```diff
@@ -0,0 +1,92 @@
+        transactionPrinter.printMostWithdrawalCategory(getMostWithdrawalCategory());
+    }
+
+    private int getAmountDeposit() {
```
저도 이부분이 고민되는 부분이었어요!
- 현준님처럼 메소드로 빼서 사용하는게 훨씬 좋은것 같으면서도
- 비슷한 코드인데 양수 음수만 판별하는데 함수를 2개두는게 애매한것같은...?
- 그래서 저는 필드에 변수를 빼서 그걸로 카운트해서 return 해주었는데 이부분에 대해서 저도 궁금합니다ㅠㅠ!

sezikim replied:
저도 고민이 되는 부분인데요..!
저도 필드로 빼서 하는 방법을 떠올렸다가,
주말에 읽은 '토끼책'에서 객체를 설계할 때 행동 위주의 설계를 하라고 나온 내용이 기억에남아.. 메소드로 빼서 구현해봤습니다.

minhee0327 commented on src/main/java/com/sezikim/bank/TransactionAnalyzer.java:
```diff
@@ -0,0 +1,92 @@
+        return transactionCountMap;
+    }
+
+    private String getMostWithdrawalCategory() {
```
아까 이래서 개발자의 소통이 중요하다 했던...ㅎㅎㅎ
- 저는 이 항목을 구할 때 토탈 금액이 가장 많이 지출된 사람을 뽑았고 그렇게하니까 Rent 가 출력되더라구요 (-4000)
- 이부분은 각자 생각한 기준대로 구한게 답일지도 모르겠기도하고...일단 책 2챕터를 다 읽고 한번 확인 또 해봐도 좋을거같아요..!

minhee0327 commented on src/main/java/com/sezikim/bank/model/Transaction.java:
```diff
@@ -0,0 +1,36 @@
+import java.time.LocalDate;
+
+public class Transaction {
+    public static final String TRANSACTION_FORMAT = "%-20.20s %-20.20s %-20.20s";
```
아까 이 부분  궁금했던 부분인데 이건 어떻게 쓰이는지 format인지 혹시 알수있을까요?? 🙌

sezikim replied:
거래내역 String 출력하면서 다음과 같이 여백(?) 맞춰서 출력하고자 했습니당..!
![스크린샷 2021-08-16 오후 9 57 04](https://user-images.githubusercontent.com/41041688/129567247-ce8148b9-a5be-441d-9bd2-83625c07ebd6.png)

minhee0327 reviewed (commented):
아까 같이 보면서 이야기 했던 부분을 위주로 보았는데요!
중간에 또 생각이 더 나면 더 리뷰해 보도록 하겠습니다 🤗
제 눈에는...잘하신거 같습니다..!(?)!!

WooSungHwan commented:
메소드 위에 어떤 역할을 하는 메소드인지 설명하는 주석을 달아보시는 건 어떤가요??

sezikim commented:
@minhee0327 @WooSungHwan 리뷰 감사합니다 👍 😄 
피드백 반영해서 PR 올리겠습니다.

sezikim committed:
refactor. 피드백 반영 및 Parsing 방안 수정
1. Category -> Content 변수 및 메소드명 변경
2. 메소드 주석 설정.
3. CSV파일 Parsing하는 방법 StringTokenizer -> split 함수로 수정

sezikim committed:
Delete libraries-with-intellij-classes.xml
